### ISSUE TITLE
Update CouchDB to version 3.4.1

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#12127]
**Impact of the new feature**
CouchDB (hence most of the WM ecosystem)

**Is your feature request related to a problem? Please describe.**
We are currently running CouchDB 3.2.2 in both central and local agent, which has been released about 2 years ago.
In order not to accumulate too much technical debt, we should look into upgrading CouchDB to the latest version, 3.4.1, released a few days ago.

In addition to keeping up with the latest versions, this version has a new JavaScript engine that is 4x faster than SpiderMonkey, reduced memory footprint, and some other interesting things. Please refer to the release notes for more information:
https://docs.couchdb.org/en/stable/whatsnew/3.4.html

**Describe the solution you'd like**
Update CouchDB used both in central services
https://github.com/dmwm/CMSKubernetes/blob/master/docker/couchdb/Dockerfile

and local wmagent
https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/wmagent-couchdb/Dockerfile

to version `3.4.1`.

Together with this upgrade, functional tests need to be performed to ensure that the correctness of the WM system.

**Describe alternatives you've considered**
An alternative on how this gets implemented, is to actually merge both Dockerfiles and have a single image that can be used both in the agent and in central services.

**Additional context**
None

**Comment by vkuznet** [comment]
We can update prometheus couchdb exporter (https://github.com/dmwm/CMSKubernetes/pull/1623) together with new CouchDB version.

**Comment by vkuznet** [comment]
The new PR: https://github.com/dmwm/CMSKubernetes/pull/1641 provides all necessary changes to upgrade CouchDB to new version and at the same time upgrade prometheus CouchDB exporter.

**Comment by amaltaro** [comment]
Thank you for providing this information, @vkuznet . My understanding with this is that upgrading CouchDB version should be fairly transparent, and there isn't a need for in-depth preparations, as 3.4.1 can read data/shards from 3.2.2 (only exception is if rollback is required).

In addition, I wonder if you read anything about database replication? Do you know if database replication works between the two different versions?

**Comment by vkuznet** [comment]
Alan, yes I don't foresee any issue with replication either. To prove that I used the following procedure:
- I used cmsweb-test10 with CouchDB 3.2.2 installed
- I used vocms0131 with CouchDB 3.4.1 installed
- I setup replication manually between cmsweb-test10 and vocms0131 via the following procedure:
```
# on source machine I created new testdb database
curl -X PUT http://uuser:password@localhost:5984/testdb

# on target node I created testdb database
curl -X PUT http://uuser:password@vocms0131.cern.ch:5984/testdb

# on local node I injected new document
curl -X POST http://user:password@localhost:5984/testdb \
     -H "Content-Type: application/json" \
     -d '{"_id": "doc1", "type": "example", "msg": "hello world}'
curl -X POST http://user:password@localhost:5984/testdb \
     -H "Content-Type: application/json" \
     -d '{"_id": "doc2", "type": "example", "msg": "hello from couch 3.2.2"}'

# on local node I setup new replication
curl -X POST http://user:password@localhost:5984/_replicate \
     -H "Content-Type: application/json" \
     -d '{"source": "http://user:password@localhost:5984/testdb", \
          "target": "http://user:password@vocms0131.cern.ch:5984/testdb"}'

# on target node I checked documents in testdb
curl http://localhost:5984/testdb/_all_docs
{"total_rows":2,"offset":0,"rows":[
{"id":"doc1","key":"doc1","value":{"rev":"1-559be9d24032a456dce21afb1d68a740"}},
{"id":"doc2","key":"doc2","value":{"rev":"1-3a9818ea831fec26458f01681205eee5"}}
]}

# check the doc2 document content on target node
curl http://localhost:5984/testdb/doc2
{"_id":"doc2","_rev":"1-3a9818ea831fec26458f01681205eee5",
 "type":"example","msg":"hello from couch 3.2.2"}
```

Therefore, from this exercise I can conclude that replication works just fine.

**Comment by vkuznet** [comment]
I performed an additional tests asked by Alan (@amaltaro):
- insert document into CouchDB 3.4.1 and then rollback the image to version 3.2.2 and verify that we can still read the document:
```
# under version 3.4.1 I injected the following document
curl -X POST http://user:pass@localhost:5984/testdb -H "Content-Type: application/json" -d '{"_id": "doc3", "type": "example", "msg": "document inserted to 3.4.1 CouchDB"}'

# then I stopped container and redeploy CouchDB 3.2.2
curl http://localhost:5984/
{"couchdb":"Welcome","version":"3.2.2","git_sha":"d5b746b7c","uuid":"ea73c6c54bbe95b2599702e0931ab737","features":["access-ready","partitioned","pluggable-storage-engines","reshard","scheduler"]}

# and now we can verify that the document `doc3` is still readable
curl http://localhost:5984/testdb/doc3
{"_id":"doc3","_rev":"1-f9b9d41f5ecb663ecc65628f6659ba92","type":"example","msg":"document inserted to 3.4.1 CouchDB"}
```
- deploy CouchDB 3.4.1 on test10 dev cluster node and configure some agent to use it for data storage. I'll update this part once it is done.

**Comment by vkuznet** [comment]
Further testing reveals the following

```
scurl -X POST -H "Content-Type: application/json" https://xxx:zzz@cmsweb-test10.cern.ch/couchdb/_replicate -d '{"source": "https://xxx:zzz@cmsweb-test10.cern.ch/couchdb/testdb", "target": "https://xxx:zzz@cmsweb-test11.cern.ch/couchdb"}'
{"error":"replication_auth_error","reason":"{session_request_failed,\"https://cmsweb-test10.cern.ch/_session\",\"userfoo\",\n    {conn_failed,\n        {error,\n            {tls_alert,\n                {decode_error,\n                    \"TLS client: In state wait_cert_cr at tls_handshake.erl:442 generated CLIENT ALERT: Fatal - Decode Error\\n handshake_decode_error\"}}}}}"}
```

And, I found the following differences between test10 and test11:
- test11
```
scurl -vk https://cmsweb-test11.cern.ch/couchdb/_session
...
*  subject: DC=ch; DC=cern; OU=computers; CN=cmsweb-test1.cern.ch
...
< server: CouchDB/3.2.2 (Erlang OTP/23)
< x-content-type-options: bla
< content-length: 102
<
* TLSv1.3 (IN), TLS app data, [no content] (0):
{"ok":true,"userCtx":{"name":null,"roles":[]},"info":{"authentication_handlers":["proxy","default"]}}
```
- test10
```
...
*  subject: DC=ch; DC=cern; OU=computers; CN=cmsweb-test1.cern.ch
...
< server: CouchDB/3.4.1 (Erlang OTP/25)
< x-content-type-options: bla
< content-length: 102
<
* TLSv1.3 (IN), TLS app data, [no content] (0):
{"ok":true,"userCtx":{"name":null,"roles":[]},"info":{"authentication_handlers":["proxy","default"]}}
```

---

Using ChatGPT I found the following (needs to be cross-checked):
- Node A: CouchDB 3.4.1 on Erlang/OTP 25 → has full TLS 1.3 support.
- Node B: CouchDB 3.2.2 on Erlang/OTP 23 → no real TLS 1.3 support (OTP 23 negotiates up to TLS 1.2; TLS 1.3 came in OTP 24).

Even though the handshake debug lines say TLSv1.3 (IN), TLS app data, [no content], that’s Wireshark/openssl labeling the message, not confirmation that Erlang 23 can actually parse it. What’s happening is: the frontend/server proposes TLS 1.3, OTP 23 tries to read it, and fails → handshake_decode_error.

To fix the problem it suggests:

1. Move the CouchDB 3.2.2 / OTP 23 node up to at least CouchDB 3.3.x with OTP 24+, ideally match the other node (3.4.1/OTP 25).
2. or restrict frontend/proxy to allow TLS 1.2
3. or, force CouchDB replication to use TLS 1.2 via changing CouchDB confiugation to use
```
[ssl]
versions = tlsv1.2
```
Based on these suggestions I think our option should be to try to upgrade first CouchDB to version 3.3.x and verify that TLS handshaking works using TLS 1.3 protocol. If it is the case then move to upgrade to version 3.4.x

**Comment by vkuznet** [comment]
New images:
- [registry.cern.ch/cmsweb/couchdb:3.5.0](http://registry.cern.ch/cmsweb/couchdb:3.5.0)
- [registry.cern.ch/cmsweb/couchdb:3.4.3](http://registry.cern.ch/cmsweb/couchdb:3.4.3)
- [registry.cern.ch/cmsweb/wmagent-couchdb:3.5.0](http://registry.cern.ch/cmsweb/wmagent-couchdb:3.5.0)
- [registry.cern.ch/cmsweb/wmagent-couchdb:3.4.3](http://registry.cern.ch/cmsweb/wmagent-couchdb:3.4.3)

**Comment by vkuznet** [comment]
> [@vkuznet](https://github.com/vkuznet) Is there any signs that Erlang is aware of this? It seems like it would be considered a serious bug with some response from the devs plus a plan to address (unless of course it represents just a version in compatibility with TLS where our tools are not up-to-date).

@klannon , based on my experience with Erlang, which has a strong track record in telecom and other reliability critical domains, I find it unlikely that the Erlang team is unaware of this behavior. More plausibly, OTP 24 introduced stricter rules for parsing X.509 certificates, and the CERN certificate chain may expose this—possibly due to encoding quirks, unusual extensions, or an empty CA list in the CertificateRequest.

Further investigation is needed to pinpoint the root cause. A practical next step would be to test with a different certificate chain—for example, one issued by Let’s Encrypt—so we can determine whether the issue is specific to the CERN CA. If alternative chains work without error, we can escalate this with the CERN certificate authority team. Otherwise, a reproducible test case using a non-CERN chain would be the ideal input to raise with the Erlang/OTP maintainers.
